teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,297 @@
+  - According to [SLSA v0.2](https://slsa.dev/provenance/v0.2), if **invocation.configSource** is not available then **buildConfig** can be used to verify information about the build. If the policy requires validating the underlying task/pipeline spec that was not fetched from version control then the information from the spec required for validation must be embedded in the provenance’s buildConfig. 
+
+## Goals
+This proposal attempts to do the following:
+
+- Identify configuration, taskSpec and task-run-time information that lead to the production of the build but is currently missing in the provenance.
+- Identify where in the provenance to host this missing information.
+
+## Out-of-scope
+- Reproducibility of complete build instructions using a **Pipeline Run**.
+- A **Resolved Inputs** field for TaskRun where "resolved inputs" is a direct encoding of reproducibility (and verified during the build pod itself via the entrypointer). In the long term, it is useful to have reproducibility metadata built into pipelines so that the changes to the TaskRun will need to consider reproducibility and chains does not have to be constantly updated whenever there is a drift. This field will require its own TEP to spec-out the details and therefore it is currently out-of-scope.
+- **Level of reproducibility** in the provenance: 
+  - **Full** -- A user can rerun the exact steps with the exact inputs and get the exact results, with all <choose your favorite hash> signatures matching for inputs and artifacts. Until we have provenance that captures the full toolchain version stack -- possibly even down to the OS kernel version(?) -- we're going to be very hard-pressed to guarantee full reproducibility in the initial implementation.
+  - **Partial** -- We can guarantee that a user can run the exact steps with the exact inputs -- but the output artifacts could change due to toolchain changes. For example, imagine all inputs are identical but there is a new compiler optimization. The output artifact is functionally identical but the signature doesn't match.
+- **Hermetic builds**
+  - Builds are only reproducible if they are [hermetic](https://slsa.dev/spec/v0.1/requirements#hermetic) as defined by SLSA. Hermetic build does not guarantee reproducibility (eg. If the build script links library1 or library2 depending on whether build process id or timestamp is a particular value then running the build again might not give the same result) but if a build is not hermetic then we cannot guarantee reproducibility. Tekton introduced hermetic builds in [TEP 0025](https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md). Since then, it has been an experimental feature. This feature needs to be pulled out of an experimental feature and [propagated to beta](https://github.com/tektoncd/pipeline/issues/3965https://github.com/tektoncd/pipeline/issues/3965) and beyond to unblock reproducible builds.
```

Comment:
> Builds are only reproducible if they are hermetic

That's a fairly strong statement. I can run reproducible builds of Go binaries with external (pinned) dependencies today.

Reply:
Hmm. I see. How about I make it a bit flexible: 

According to SLSA L4 requirements, builds must be [hermetic](https://slsa.dev/spec/v0.1/requirements#hermetic). But if you say that L4 is going away in 1.0 version of the spec then should I just remove that line about Hermeticity? Keep everything else from the next sentence?

Reply:
I think it's useful to call out that hermetic builds are out of scope here, while also being helpful to capturing the inputs to the build. So I'd drop the L4 reference and soften the first sentence.

Reply:
Sounds good! Done.